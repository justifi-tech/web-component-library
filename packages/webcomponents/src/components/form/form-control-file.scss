@import 'reboot';
@import 'forms';
@import 'badge';

:host {
  @include host-base-styles;
  display: block;
}

input[type='file'][disabled] {
  color: var(--jfi-form-control-disabled-color);
  background-color: var(--jfi-form-control-disabled-background-color);
  opacity: 0.5;
  cursor: not-allowed;
}

.input-group-text {
  background-color: var(--jfi-form-control-background-color);
  border-color: var(--jfi-form-control-border-color);
  color: var(--jfi-form-control-color);
}

.file {
  border-top-right-radius: 0 !important;
  border-bottom-right-radius: 0 !important;
}
