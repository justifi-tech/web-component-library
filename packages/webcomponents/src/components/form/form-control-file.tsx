import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
} from '@stencil/core';

@Component({
  tag: 'form-control-file',
  styleUrl: 'form-control-file.scss',
  shadow: true,
})
export class FileInput {
  @Element() el: HTMLElement;
  @Prop() label: string;
  @Prop() name: any;
  @Prop() error: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;
  @State() input: string;
  @State() file: File;
  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  updateInput(newValue: any) {
    const inputElement = this.el.shadowRoot.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  handleFileChange(event: any) {
    this.file = event.target.files[0];
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          type="file"
          name={this.name}
          part={`input ${this.error ? 'input-invalid ' : ''}${this.disabled ? ' input-disabled' : ''}`}
          class={this.error ? 'form-control is-invalid' : 'form-control'}
          disabled={this.disabled}
          onChange={(event: any) => this.handleFileChange(event)}
          onInput={(event: any) => this.handleFormControlInput(event)}
          onBlur={() => this.formControlBlur.emit()}
        />
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
