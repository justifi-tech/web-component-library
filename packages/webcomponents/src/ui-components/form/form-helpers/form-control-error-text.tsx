import { h } from '@stencil/core';

export const FormControlErrorText = (props: { errorText: string, name: string }) => {
  const { errorText, name } = props;

  if (!errorText) {
    return null;
  }

  return (
    <small class='form-text text-danger' id={`form-error-text-${name}`} part="input-error-text" >
      {errorText}
    </small>
  );
}
