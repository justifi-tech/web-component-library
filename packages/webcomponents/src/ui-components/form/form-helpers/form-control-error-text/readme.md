# form-control-error-text



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `errorText` | `error-text` |             | `string` | `undefined` |
| `name`      | `name`       |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-checkbox](../..)
 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-monetary](../..)
 - [form-control-number](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [form-control-textarea](../..)
 - [justifi-dispute-reason](../../../../components/dispute-management/dispute-response)

### Graph
```mermaid
graph TD;
  form-control-checkbox --> form-control-error-text
  form-control-date --> form-control-error-text
  form-control-file --> form-control-error-text
  form-control-file-v2 --> form-control-error-text
  form-control-monetary --> form-control-error-text
  form-control-number --> form-control-error-text
  form-control-number-masked --> form-control-error-text
  form-control-select --> form-control-error-text
  form-control-text --> form-control-error-text
  form-control-textarea --> form-control-error-text
  justifi-dispute-reason --> form-control-error-text
  style form-control-error-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
