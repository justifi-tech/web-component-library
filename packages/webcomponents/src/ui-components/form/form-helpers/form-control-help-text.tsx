import { h } from '@stencil/core';

export const FormControlHelpText = (props: { helpText: string, name: string }) => {
  const { helpText, name } = props;

  if (!helpText) {
    return null;
  }

  return (
    <small class='form-text text-muted' id={`form-help-text-${name}`} part="input-help-text">
      {helpText}
    </small>
  );
}
