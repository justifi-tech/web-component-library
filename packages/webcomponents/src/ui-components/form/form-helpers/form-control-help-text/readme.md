# form-control-help-text



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type     | Default     |
| ---------- | ----------- | ----------- | -------- | ----------- |
| `helpText` | `help-text` |             | `string` | `undefined` |
| `name`     | `name`      |             | `string` | `undefined` |


## Dependencies

### Used by

 - [bank-account-form-inputs](../../../../components/business-forms/payment-provisioning/bank-account)
 - [bank-account-form-inputs-canada](../../../../components/business-forms/payment-provisioning/bank-account)
 - [form-control-radio](../..)
 - [form-control-textarea](../..)

### Graph
```mermaid
graph TD;
  bank-account-form-inputs --> form-control-help-text
  bank-account-form-inputs-canada --> form-control-help-text
  form-control-radio --> form-control-help-text
  form-control-textarea --> form-control-help-text
  style form-control-help-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
