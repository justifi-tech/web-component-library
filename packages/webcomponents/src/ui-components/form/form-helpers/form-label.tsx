import { h } from '@stencil/core';

export const FormLabel = ({ htmlFor, label, helpText }) => {
  return (
    <label
      part="label"
      class="form-label d-flex gap-2 align-items-center"
      htmlFor={htmlFor}
    >
      {label}
      <form-control-tooltip helpText={helpText} />
    </label>);
}
