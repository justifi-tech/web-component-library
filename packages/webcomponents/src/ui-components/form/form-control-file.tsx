import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
} from '@stencil/core';
import { EntityDocumentType, FileSelectEvent } from '../../api/Document';
import { FormControlErrorText } from '../../ui-components';
import { FormLabel } from './form-helpers/form-label';

@Component({
  tag: 'form-control-file',
})
export class FileInput {
  @Element() el: HTMLElement;
  fileInput: HTMLInputElement;

  @State() files: File[];
  @State() fileString: string;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() multiple?: boolean;
  @Prop() documentType: EntityDocumentType;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;
  @Event() fileSelected: EventEmitter<FileSelectEvent>;

  componentDidLoad() {
    this.fileInput = this.el.querySelector('input');
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit({ name, value: target.value });
  }

  changeHandler = () => {
    const inputFileList = this.fileInput.files;
    if (inputFileList) {
      this.fileSelected.emit({
        fileList: inputFileList,
        document_type: this.documentType
      });
    }
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <FormLabel
            htmlFor={this.name}
            label={this.label}
            helpText={this.helpText}
          />
          <input
            ref={(el) => this.fileInput = el}
            type="file"
            name={this.name}
            part={`input ${this.errorText ? "input-invalid " : ""}${this.disabled ? "input-disabled" : ""}`}
            class={this.errorText ? "form-control is-invalid" : "form-control"}
            multiple={this.multiple}
            disabled={this.disabled}
            onChange={this.changeHandler}
            onInput={this.handleFormControlInput}
            onBlur={() => this.formControlBlur.emit()}
          />
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
