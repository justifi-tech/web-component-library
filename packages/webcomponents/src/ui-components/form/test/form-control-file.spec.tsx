import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { FileInput } from '../form-control-file';
import { FormControlErrorText } from '../form-helpers/form-control-error-text';
import { FormControlHelpText } from '../form-helpers/form-control-help-text';

describe('form-control-file', () => {
  const components = [FileInput, FormControlErrorText, FormControlHelpText];
  const mockInputHandler = jest.fn();

  it('Renders with default props', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-file label='Profile Picture' name='profilePicture'></form-control-file>,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.label).toBe('Profile Picture');
    expect(page.rootInstance.name).toBe('profilePicture');
  });

  it('Renders with all props provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Resume'
          name='resume'
          errorText='Invalid file type'
          helpText='Upload your resume in PDF format'
          disabled
          inputHandler={mockInputHandler}
        >
        </form-control-file>
    });

    expect(page.root).toMatchSnapshot();
  });

  it('Handles user input correctly', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Government ID'
          name='governmentId'
          inputHandler={mockInputHandler}
        >
        </form-control-file>
    });

    const inputElement = page.root.querySelector('input');
    const testValue = 'govermnent-id.jpg';

    inputElement.value = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.value).toBe(testValue);
  });

  it('Emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Government ID'
          name='governmentId'
          inputHandler={mockInputHandler}
        >
        </form-control-file>
    });

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.value = 'govermnent-id.jpg';

    inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    expect(inputEventSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: expect.objectContaining({
        name: 'governmentId',
        value: 'govermnent-id.jpg'
      })
    }));
  });

  it('Emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Government ID'
          name='governmentId'
          inputHandler={mockInputHandler}
        >
        </form-control-file>
    });

    const blurEventSpy = jest.fn();
    page.root.addEventListener('formControlBlur', blurEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.dispatchEvent(new Event('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('Disables input when disabled prop is true', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Government ID'
          name='governmentId'
          disabled
        >
        </form-control-file>
    });

    const inputElement = page.root.querySelector('input');
    expect(inputElement.disabled).toBeTruthy();
  });

  it('Shows help text when helpText prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Government ID'
          name='governmentId'
          inputHandler={mockInputHandler}
          helpText='Select a file to upload.'
        >
        </form-control-file>
    });

    const helpTextComponent = page.root.querySelector('#form-help-text-governmentId');
    expect(helpTextComponent).not.toBeNull();

    expect(helpTextComponent.textContent).toBe('Select a file to upload.');
  });

  it('Shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-file
          label='Government ID'
          name='governmentId'
          inputHandler={mockInputHandler}
          errorText='This field is required.'
        >
        </form-control-file>
    });

    const errorTextComponent = page.root.querySelector('#form-error-text-governmentId');
    expect(errorTextComponent).not.toBeNull();

    expect(errorTextComponent.textContent).toBe('This field is required.');

    const inputElement = page.root.querySelector('input');
    expect(inputElement.classList.contains('is-invalid')).toBe(true);
  });
});
