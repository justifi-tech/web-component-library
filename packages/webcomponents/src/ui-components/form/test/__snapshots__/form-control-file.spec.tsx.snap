// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-file Renders with all props provided 1`] = `
<form-control-file exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <label class="form-label" htmlfor="resume" part="label">
      Resume
    </label>
    <input class="form-control is-invalid" disabled="" name="resume" part="input input-invalid " type="file">
    <small class="form-text text-muted" id="form-help-text-resume" part="input-help-text">
      Upload your resume in PDF format
    </small>
    <small class="form-text text-danger" id="form-error-text-resume" part="input-error-text">
      Invalid file type
    </small>
  </div>
</form-control-file>
`;

exports[`form-control-file Renders with default props 1`] = `
<form-control-file exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <label class="form-label" htmlfor="profilePicture" part="label">
      Profile Picture
    </label>
    <input class="form-control" name="profilePicture" part="input " type="file">
  </div>
</form-control-file>
`;
